File: js/313_SuperUglyNumber.js
/**
 * @param {number} n
 * @param {number[]} primes
 * @return {number}
 */
var nthSuperUglyNumber = function(n, primes) {
    var len = primes.length;
    var mark = [];
    for(var m = 0; m < len; ++m){
        mark.push(0);
    }
    var dp = [];
    dp.push(1);
    for(var i = 1; i < n; ++i){
        var min = Number.MAX_SAFE_INTEGER;
        for(var j = 0; j < len; ++j){
            min = Math.min(min, dp[mark[j]]*primes[j]);
        }
        for(var k = 0; k < len; ++k){
            if(min == dp[mark[k]]*primes[k]) mark[k]++;
        }
        dp.push(min);
    }
    return dp[n-1];
};

File: cpp/313_SuperUglyNumber.cpp
class Solution {
public:
    int nthSuperUglyNumber(int n, vector<int>& primes) {
        vector<int> mark(primes.size(),0);
        int len = primes.size();
        vector<int> dp;
        for(int i = 0; i < n; ++i){
            if( i == 0){
                dp.push_back(1);
                continue;
            }
            int min = INT_MAX;
            for(int j = 0; j < len; ++j){
                min = std::min(dp[mark[j]] * primes[j], min);
            }
            for(int j = 0; j < len; ++j){
                if(min == dp[mark[j]] * primes[j]) mark[j]++;
            }
            dp.push_back(min);
        }
        return dp[n-1];
    }
};